Semigroup-Monoid.md

```diff
@@ -0,0 +1,318 @@
+^ You can use it as soon as you add it to your build.gradle file.
+
+---
+
+# Semigroup
+
+`Semigroup` is a __Type Class__, so it defines a given behavior.
+
+`Monoid` inherit its combinators. 
+
+```
+Semigroup
+
+Monoid
+```
+
+^ Semigroup is defined as a Type Class. In pure Functional Programming, type classes define behaviors.
+^ Other very well known type classes are Monoids, and we will talk about it later in this same video since they are intimately related.
+
+---
+
+# Semigroup :: Polymorphism
+
+```kotlin
+interface Semigroup<A> {
+    //...
+}
+```
+
+- Semigroup is __parametric over a type constructor `A`__. This allows us to build abstract functions over the behaviours that Semigroup defines and forget about the concrete types that A may refer to.
+
+- We call this __ad-hoc polymorphism__, the ability to write polymorphic programs that can be defined in generic terms.
+
+^ Semigroup is parametric over a generic type A. Here, A stands for a data type, also called "Type Constructor".
+^ Thanks to this, we can declare generic functions encoded on top of the behaviours provided by Semigroup and forget all about the concrete data type used for A.
+^ We called this "Ad-Hoc polymorphism" and that's how type classes allow us to encode completely generic and polymorphic programs that can work over many different data types.
+
+---
+
+# Semigroup :: combine
+
+`Semigroup` abstracts the ability to __combine__ two values of the same type `A`.
+
+In other words, it provides a combining function for the `A` type:
+
+```kotlin
+fun A.combine(b: A): A
+```
+
+^ The most important behaviour provided by Semigroup is the combine function.
+^ With combine, we can combine/concat/join two A values.
+^ Semigroup defines on a type A a combination function which takes as input two elements of A to return an element of the same type A.
+
+---
+
+# Semigroup :: Data types
+
+Any type constructor whose contents can be combined can provide an instance of `Semigroup`.
+
+- `Option<A>`
+- `NonEmptyList<A>`
+- `ListK<A>`
+- `SequenceK<A>`
+- `SetK<A>`
+- `MapK<A>`
+- `SortedMapK<A>`
+- `Either<A>`
+- `Try<A>`
+
+Arrow also offers Semigroup instances to native Kotlin `String`, `Byte`, `Double`, `Int`, `Long`, `Short` and `Float` types.
+
+^ In order for a data type content to be “combinable”, we need its type constructor to be able to provide an instance of a Semigroup. Some data types examples on Arrow that offer exactly that would be: Option, NonEmptyList, ListK, SequenceK, SetK...
+
+---
+
+# Semigroup :: combine
+
+In this example we are showing the main characteristic about `Semigroup`, the `combine` method.
+As you can see here, we can apply it for both a `Double` and a `String` type, but remember you can use it with many other types.
+
+
+```kotlin
+with(Double.semigroup()){
+    4.5.combine(7.0).combine(2.5)
+}
+//14.0
+```
+
+```kotlin
+with(String.semigroup()){
+    "This is ".combine("functional ") + ("programming.")
```

Comment:
May be clear to use combine multiple times instead of +

Reply:
Ok, done